src/main/java/org/mockito/internal/util/KotlinInlineClassUtil.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal.util;
+
+import org.mockito.internal.stubbing.BaseStubbing;
+import org.mockito.internal.stubbing.OngoingStubbingImpl;
+import org.mockito.invocation.MatchableInvocation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class KotlinInlineClassUtil {
+    // When mocking function, returning inline class, its return type is
+    // underlying type.
+    // So, `thenReturn` calls fails, because of non-compatible types.
+    public static boolean isInlineClassWithAssignableUnderlyingType(
+            Class<?> inlineClass, Class<?> underlyingType) {
+        try {
+            // All inline classes have 'box-impl' method, which accepts
+            // underlying type and returns inline class.
```

Comment:
In this case, yes I think we should ideally wait for Kotlin 1.5 to ship. Of course we need to be compatible with previous Kotlin versions. Therefore, we could check if the annotation exists on the class and return that?

Since `Class.getAnnotation` takes a `Class`, the annotation would be on the classpath. Therefore, we can optimize `unboxUnderlyingValueIfNeeded` to only run when the annotation is available. E.g. we load the class based on its string name in a `static {}` block into a field. If that throws a `ClassNotFoundException`, the `Class<?>` field remains `null` and we can short-circuit `unboxUnderlyingValueIfNeeded` to bail if the class is `null`. We use a similar function for some JVM-internals in https://github.com/mockito/mockito/blob/80374488162380d006b8ae6da915d940cbf68c7c/src/main/java/org/mockito/internal/exceptions/stacktrace/StackTraceFilter.java#L23-L39